lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2256,6 +2256,120 @@ ClangModuleUnit *ClangImporter::Implementation::getClangModuleForDecl(
+  // In some situations, the ClangImporter is invoked in a reentrant manner,
+  // resulting in a diagnosticTarget already being set. In such a case, we
+  // keep the outer diagnostic target.
+  if (!diagnosticTarget) {
+    diagnosticTarget = target;
+    return true;
+  }
+  return false;
+}
+
+SwiftLookupTable::SingleEntry
+ClangImporter::Implementation::getDiagnosticTarget() {
+  return diagnosticTarget;
+}
+
+void ClangImporter::Implementation::resetDiagnosticTarget(bool shouldReset) {
+  if (shouldReset) {
+    assert(diagnosticTarget);
+    diagnosticTarget = nullptr;
+  }
+}
+
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &targetCandidate) const {
+  return diagnosticTarget == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::NamedDecl *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::NamedDecl *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::ModuleMacro *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::ModuleMacro *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::MacroInfo *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::MacroInfo *>() == targetCandidate;
+}
+
+bool ClangImporter::Implementation::setEagerImportActive() {
+  bool shouldReset = !eagerImportActive;
+  eagerImportActive = true;
+  return shouldReset;
+}
+
+void ClangImporter::Implementation::resetEagerImportActive(bool shouldReset) {
+  if (shouldReset) {
+    eagerImportActive = false;
+  }
+}
+
+bool ClangImporter::Implementation::getEagerImportActive() {
+  return eagerImportActive;
+}
+
+void ClangImporter::Implementation::reportNoteConsumingDiagnosticError(
+    const clang::NamedDecl *target, Diagnostic &&error,
+    const clang::SourceLocation &loc) {
+  if (SwiftContext.LangOpts.EnableExperimentalClangImporterDiagnostics &&
+      isDiagnosticTarget(target)) {
+    diagnose(convertSourceLocation(loc), error);
+    for (auto noteIt = pendingErrorNotes.rbegin();
+         noteIt != pendingErrorNotes.rend(); ++noteIt) {
+      if (!noteIt->diag.getLoc()) {
+        noteIt->diag.setLoc(convertSourceLocation(loc));
+      }
+      if (isDiagnosticTarget(noteIt->target)) {
+        diagnose(noteIt->diag.getLoc(), noteIt->diag);
+      }
+    }
+  }
+  pendingErrorNotes.clear();
+}
+
+void ClangImporter::Implementation::addPendingErrorNote(Diagnostic &&note) {
+  addPendingErrorNote(std::move(note), clang::SourceLocation());
+}
+
+void ClangImporter::Implementation::addPendingErrorNote(
+    Diagnostic &&note, const clang::SourceLocation &loc) {
+  if (SwiftContext.LangOpts.EnableExperimentalClangImporterDiagnostics &&
+      getDiagnosticTarget()) {
+    for (auto &pendingErrorNote : pendingErrorNotes) {
+      if (!pendingErrorNote.diag.getLoc() &&
+          diagnosticTarget == getDiagnosticTarget()) {
```

Comment:
What is this comparison doing? Is `diagnosticTarget` ever not the same as `getDiagnosticTarget`?

Reply:
Ooops, this is a mistake, I meant `pendingErrorNote.target == getDiagnosticTarget()`. The idea is only fix up notes targeted at the current target, but it is not vital, the remaining notes will be discarded as it stands anyways.